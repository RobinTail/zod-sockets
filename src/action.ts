import { init, last } from "ramda";
import { z } from "zod";
import { ActionNoAckDef, ActionWithAckDef } from "./actions-factory";
import { EmissionMap } from "./emission";
import { ActionContext, ClientContext, Handler } from "./handler";
import { Namespace, Namespaces, rootNS } from "./namespaces";

export abstract class AbstractAction {
  public abstract getEvent(): string;
  public abstract getNamespace(): string;
  public abstract execute(
    params: {
      event: string;
      params: unknown[];
    } & ClientContext<EmissionMap>,
  ): Promise<void>;
  public abstract getSchema(variant: "input"): z.AnyZodTuple;
  public abstract getSchema(variant: "output"): z.AnyZodTuple | undefined;
}

export class Action<
  IN extends z.AnyZodTuple,
  OUT extends z.AnyZodTuple,
> extends AbstractAction {
  readonly #event: string;
  readonly #namespace: string;
  readonly #inputSchema: IN;
  readonly #outputSchema: OUT | undefined;
  readonly #handler: Handler<
    ActionContext<z.output<IN>, EmissionMap>,
    z.input<OUT> | void
  >;

  public constructor(
    action:
      | ActionWithAckDef<IN, OUT, Namespaces<Namespace<EmissionMap>>, string>
      | ActionNoAckDef<IN, Namespaces<Namespace<EmissionMap>>, string>,
  ) {
    super();
    this.#event = action.event;
    this.#namespace = action.ns || rootNS;
    this.#inputSchema = action.input;
    this.#outputSchema = "output" in action ? action.output : undefined;
    this.#handler = action.handler;
  }

  public override getEvent(): string {
    return this.#event;
  }

  public override getNamespace(): string {
    return this.#namespace;
  }

  public override getSchema(variant: "input"): IN;
  public override getSchema(variant: "output"): OUT | undefined;
  public override getSchema(variant: "input" | "output") {
    return variant === "input" ? this.#inputSchema : this.#outputSchema;
  }

  /** @throws z.ZodError */
  #parseInput(params: unknown[]) {
    const payload = this.#outputSchema ? init(params) : params;
    return this.#inputSchema.parse(payload);
  }

  /** @throws z.ZodError */
  #parseAckCb(params: unknown[]) {
    if (!this.#outputSchema) {
      return undefined;
    }
    return z.function(this.#outputSchema, z.void()).parse(last(params));
  }

  /** @throws z.ZodError */
  #parseOutput(output: z.input<OUT> | void) {
    if (!this.#outputSchema) {
      return;
    }
    return this.#outputSchema.parse(output) as z.output<OUT>;
  }

  public override async execute({
    event,
    params,
    logger,
    ...rest
  }: {
    event: string;
    params: unknown[];
  } & ClientContext<EmissionMap>): Promise<void> {
    try {
      const input = this.#parseInput(params);
      logger.debug(
        `parsed input (${this.#outputSchema ? "excl." : "no"} ack)`,
        input,
      );
      const ack = this.#parseAckCb(params);
      const output = await this.#handler({ input, logger, ...rest });
      const response = this.#parseOutput(output);
      if (ack && response) {
        logger.debug("parsed output", response);
        ack(...response);
      }
    } catch (error) {
      logger.error(`${event} handling error`, error);
    }
  }
}
