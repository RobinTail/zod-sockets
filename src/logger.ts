/**
 * @desc Using module augmentation approach you can set the type of the actual logger used
 * @example declare module "express-zod-api" { interface LoggerOverrides extends winston.Logger {} }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
export interface LoggerOverrides {}

/** @desc You can use any logger compatible with this type. */
export type AbstractLogger = Record<
  "info" | "debug" | "warn" | "error",
  (message: string, meta?: any) => any
> &
  LoggerOverrides;
