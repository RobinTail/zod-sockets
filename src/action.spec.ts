import { Socket } from "socket.io";
import { z } from "zod";
import { AbstractAction, Action } from "./action";
import { AbstractLogger } from "./logger";

describe("Action", () => {
  const simpleHandler = vi.fn();
  const simpleAction = new Action({
    event: "simple",
    input: z.tuple([z.string()]),
    handler: simpleHandler,
  });
  const ackHandler = vi.fn(async (): Promise<[number]> => [123]);
  const ackAction = new Action({
    event: "ackOne",
    ns: "test",
    input: z.tuple([z.string()]),
    output: z.tuple([z.number()]),
    handler: ackHandler,
  });

  describe("constructor", () => {
    test("should create inheritor of AbstractAction", () => {
      expect(simpleAction).toBeInstanceOf(Action);
      expect(simpleAction).toBeInstanceOf(AbstractAction);
    });
  });

  describe(".event", () => {
    test("should be the event name", () => {
      expect(simpleAction.event).toBe("simple");
      expect(ackAction.event).toBe("ackOne");
    });
  });

  describe(".namespace", () => {
    test("should be the namespace", () => {
      expect(simpleAction.namespace).toBe("/");
      expect(ackAction.namespace).toBe("test");
    });
  });

  describe.each<keyof typeof ackAction>(["inputSchema", "outputSchema"])(
    ".%s",
    (prop) => {
      test("should be the schema", () => {
        expect(ackAction[prop]).toMatchSnapshot();
      });
    },
  );

  describe("execute()", () => {
    const loggerMock = {
      error: vi.fn(),
      debug: vi.fn(),
    };

    const commons = {
      withRooms: vi.fn(),
      all: {
        getClients: vi.fn(),
        getRooms: vi.fn(),
        broadcast: vi.fn(),
      },
      client: {
        id: "ID",
        handshake: { auth: {} } as Socket["handshake"],
        emit: vi.fn(),
        broadcast: vi.fn(),
        getRooms: vi.fn(),
        isConnected: vi.fn(),
        getRequest: vi.fn(),
        getData: vi.fn(),
        setData: vi.fn(),
        join: vi.fn(),
        leave: vi.fn(),
      },
    };

    test("should handle simple action", async () => {
      await simpleAction.execute({
        ...commons,
        logger: loggerMock as unknown as AbstractLogger,
        params: ["some"],
      });
      expect(loggerMock.error).not.toHaveBeenCalled();
      expect(simpleHandler).toHaveBeenLastCalledWith({
        ...commons,
        input: ["some"],
        logger: loggerMock,
      });
    });

    test("should handle action with ack", async () => {
      const ackMock = vi.fn();
      await ackAction.execute({
        ...commons,
        logger: loggerMock as unknown as AbstractLogger,
        params: ["some", ackMock],
      });
      expect(loggerMock.error).not.toHaveBeenCalled();
      expect(ackHandler).toHaveBeenLastCalledWith({
        ...commons,
        input: ["some"],
        logger: loggerMock,
      });
      expect(ackMock).toHaveBeenLastCalledWith(123); // from ackHandler
    });

    test("should throw input parsing error", async () => {
      await expect(
        simpleAction.execute({
          ...commons,
          logger: loggerMock as unknown as AbstractLogger,
          params: [], // first one missing
        }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    test.each(["not cb", vi.fn()])(
      "should throw acknowledgment related errors %#",
      async (ack) => {
        if (typeof ack === "function") {
          ackHandler.mockImplementationOnce(async () => [
            "not number" as unknown as number,
          ]);
        }
        await expect(
          ackAction.execute({
            ...commons,
            logger: loggerMock as unknown as AbstractLogger,
            params: ["test", ack],
          }),
        ).rejects.toThrowErrorMatchingSnapshot();
      },
    );
  });
});
