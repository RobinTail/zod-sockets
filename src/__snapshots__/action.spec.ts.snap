// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Action > execute() > should throw acknowledgment related errors 0 1`] = `
InputValidationError({
  "message": "1: Expected function, received string",
  "originalError": ZodError({
    "issues": [
      {
        "code": "invalid_type",
        "expected": "function",
        "message": "Expected function, received string",
        "path": [
          1,
        ],
        "received": "string",
      },
    ],
  }),
})
`;

exports[`Action > execute() > should throw acknowledgment related errors 1 1`] = `
OutputValidationError({
  "message": "0: Expected number, received string",
  "originalError": ZodError({
    "issues": [
      {
        "code": "invalid_type",
        "expected": "number",
        "message": "Expected number, received string",
        "path": [
          0,
        ],
        "received": "string",
      },
    ],
  }),
})
`;

exports[`Action > execute() > should throw input parsing error 1`] = `
InputValidationError({
  "message": "Array must contain at least 1 element(s)",
  "originalError": ZodError({
    "issues": [
      {
        "code": "too_small",
        "exact": false,
        "inclusive": true,
        "message": "Array must contain at least 1 element(s)",
        "minimum": 1,
        "path": [],
        "type": "array",
      },
    ],
  }),
})
`;
